File: Python/8-Graphs(BFS-DFS)/Algorithms/7-Union-Find(DisjointSetUnion-DSU)/Union-Find(DisjointSetUnion-DSU)Pro.py
"""
Union-Find (Disjoint Set Union - DSU) in Python with 6 Real-World Project Implementations

This module provides implementations of the Disjoint Set Union (DSU) data structure,
also known as Union-Find, along with examples demonstrating its application in
various real-world scenarios.

DSU is a data structure that keeps track of a set of elements partitioned into a
number of disjoint (non-overlapping) subsets. It supports two main operations:

- find(element): Determines which subset a particular element is in.  This can be
  used to determine if two elements are in the same subset.
- union(set1, set2): Joins two subsets into a single subset.

Implementations:

1.  Basic Implementation (No Optimizations):  A straightforward implementation
    using a list/array as the parent array, without any path compression or
    union by rank optimizations.  This is the easiest to understand but has
    the worst-case time complexity.

2.  Path Compression:  Improves the `find` operation by making each visited node
    point directly to the root of the set. This significantly reduces the
    height of the trees and improves performance.

3.  Union by Rank:  Improves the `union` operation by attaching the shorter
    tree to the root of the taller tree.  The rank of a tree is an upper bound
    on its height.

4.  Path Compression + Union by Rank: Combines both optimizations for maximum
    performance.  This is the most commonly used and recommended approach.

5.  DSU with List of Sets:  Instead of a parent array, this implementation uses
    a dictionary where each key is a representative element of a set, and the
    value is a set containing all elements in that subset.  This provides
    more intuitive access to the elements within each set.

6.  Iterative Path Compression: A non-recursive version of path compression,
    useful in environments with limited recursion depth.

Real-World Project Implementations:

1.  Network Connectivity: Determining if two computers are in the same network.
2.  Social Network Connectivity: Checking if two users are connected in a social network.
3.  Image Processing (Segmentation): Grouping pixels into regions.
4.  Kruskal's Algorithm (Minimum Spanning Tree):  Efficiently determining if adding
    an edge would create a cycle.
5.  Circuit Network Analysis: Identifying connected components in an electrical circuit.
6.  Maze Generation: Creating random mazes by ensuring walls are removed without
    creating cycles.
"""

from typing import List, Dict, Set, Tuple


class DSU_Basic:
    """
    Basic Disjoint Set Union implementation with no optimizations.
    """

    def __init__(self, n: int):
        """
        Initializes the DSU data structure.

        Args:
            n: The number of elements.
        """
        self.parent = list(range(n))  # Parent array: initially, each element is its own parent

    def find(self, i: int) -> int:
        """
        Finds the representative (root) of the set that element i belongs to.

        Args:
            i: The element to find the set for.

        Returns:
            The representative (root) of the set containing i.
        """
        if self.parent[i] == i:
            return i  # If i is its own parent, it's the root
        return self.find(self.parent[i])  # Recursively find the parent's parent

    def union(self, i: int, j: int) -> None:
        """
        Merges the sets containing elements i and j.

        Args:
            i: The first element.
            j: The second element.
        """
        root_i = self.find(i)  # Find the root of the set containing i
        root_j = self.find(j)  # Find the root of the set containing j
        if root_i != root_j:
            self.parent[root_i] = root_j  # Make one root the parent of the other


class DSU_PathCompression:
    """
    DSU implementation with path compression optimization.
    """

    def __init__(self, n: int):
        """
        Initializes the DSU data structure.

        Args:
            n: The number of elements.
        """
        self.parent = list(range(n))

    def find(self, i: int) -> int:
        """
        Finds the representative (root) of the set that element i belongs to,
        with path compression.

        Args:
            i: The element to find the set for.

        Returns:
            The representative (root) of the set containing i.
        """
        if self.parent[i] == i:
            return i
        self.parent[i] = self.find(self.parent[i])  # Path compression: make i point to root
        return self.parent[i]  # Return the root

    def union(self, i: int, j: int) -> None:
        """
        Merges the sets containing elements i and j.

        Args:
            i: The first element.
            j: The second element.
        """
        root_i = self.find(i)
        root_j = self.find(j)
        if root_i != root_j:
            self.parent[root_i] = root_j


class DSU_UnionByRank:
    """
    DSU implementation with union by rank optimization.
    """

    def __init__(self, n: int):
        """
        Initializes the DSU data structure.

        Args:
            n: The number of elements.
        """
        self.parent = list(range(n))
        self.rank = [0] * n  # Rank array: stores the rank of each root

    def find(self, i: int) -> int:
        """
        Finds the representative (root) of the set that element i belongs to.

        Args:
            i: The element to find the set for.

        Returns:
            The representative (root) of the set containing i.
        """
        if self.parent[i] == i:
            return i
        return self.find(self.parent[i])

    def union(self, i: int, j: int) -> None:
        """
        Merges the sets containing elements i and j, using union by rank.

        Args:
            i: The first element.
            j: The second element.
        """
        root_i = self.find(i)
        root_j = self.find(j)
        if root_i != root_j:
            if self.rank[root_i] < self.rank[root_j]:  # Attach shorter tree to taller
                self.parent[root_i] = root_j
            elif self.rank[root_i] > self.rank[root_j]:
                self.parent[root_j] = root_i
            else:
                self.parent[root_j] = root_i  # If ranks are equal, increment rank of one
                self.rank[root_i] += 1



class DSU_PathCompressionAndRank:
    """
    DSU implementation with both path compression and union by rank optimizations.
    """

    def __init__(self, n: int):
        """
        Initializes the DSU data structure.

        Args:
            n: The number of elements.
        """
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, i: int) -> int:
        """
        Finds the representative (root) of the set that element i belongs to,
        with path compression.

        Args:
            i: The element to find the set for.

        Returns:
            The representative (root) of the set containing i.
        """
        if self.parent[i] == i:
            return i
        self.parent[i] = self.find(self.parent[i])
        return self.parent[i]

    def union(self, i: int, j: int) -> None:
        """
        Merges the sets containing elements i and j, using union by rank.

        Args:
            i: The first element.
            j: The second element.
        """
        root_i = self.find(i)
        root_j = self.find(j)
        if root_i != root_j:
            if self.rank[root_i] < self.rank[root_j]:
                self.parent[root_i] = root_j
            elif self.rank[root_i] > self.rank[root_j]:
                self.parent[root_j] = root_i
            else:
                self.parent[root_j] = root_i
                self.rank[root_i] += 1



class DSU_ListOfSets:
    """
    DSU implementation using a list of sets.  This provides more direct access
    to the elements within each disjoint set.
    """

    def __init__(self, n: int):
        """
        Initializes the DSU data structure.

        Args:
            n: The number of elements.
        """
        self.sets: Dict[int, Set[int]] = {i: {i} for i in range(n)}  # {root: {elements}}
        self.element_to_root: Dict[int, int] = {i: i for i in range(n)} # {element: root}

    def find(self, i: int) -> int:
        """
        Finds the representative (root) of the set that element i belongs to.

        Args:
            i: The element to find the set for.

        Returns:
            The representative (root) of the set containing i.
        """
        return self.element_to_root[i]

    def union(self, i: int, j: int) -> None:
        """
        Merges the sets containing elements i and j.

        Args:
            i: The first element.
            j: The second element.
        """
        root_i = self.find(i)
        root_j = self.find(j)

        if root_i != root_j:
            if len(self.sets[root_i]) < len(self.sets[root_j]):
                smaller_root, larger_root = root_i, root_j
            else:
                smaller_root, larger_root = root_j, root_i

            # Merge the smaller set into the larger set
            self.sets[larger_root].update(self.sets[smaller_root])
            del self.sets[smaller_root]

            # Update the element_to_root mapping for the merged elements
            for element in self.sets[larger_root]:
                self.element_to_root[element] = larger_root

    def get_sets(self) -> Dict[int, Set[int]]:
        """
        Returns the current sets.

        Returns:
            A dictionary representing the sets.  Keys are root elements,
            values are the sets of elements.
        """
        return self.sets



class DSU_IterativePathCompression:
    """
    DSU implementation with iterative path compression.  This avoids recursion.
    """

    def __init__(self, n: int):
        """
        Initializes the DSU data structure.

        Args:
            n: The number of elements.
        """
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, i: int) -> int:
        """
        Finds the representative (root) of the set that element i belongs to,
        using iterative path compression.

        Args:
            i: The element to find the set for.

        Returns:
            The representative (root) of the set containing i.
        """
        path = []
        while self.parent[i] != i:
            path.append(i)
            i = self.parent[i]
        root = i  # Found the root

        # Iterative path compression: update parents along the path
        for node in path:
            self.parent[node] = root
        return root

    def union(self, i: int, j: int) -> None:
        """
        Merges the sets containing elements i and j, using union by rank.

        Args:
            i: The first element.
            j: The second element.
        """
        root_i = self.find(i)
        root_j = self.find(j)
        if root_i != root_j:
            if self.rank[root_i] < self.rank[root_j]:
                self.parent[root_i] = root_j
            elif self.rank[root_i] > self.rank[root_j]:
                self.parent[root_j] = root_i
            else:
                self.parent[root_j] = root_i
                self.rank[root_i] += 1



def test_dsu_implementations():
    """
    Tests the various DSU implementations with example usage and real-world scenarios.
    """
    num_elements = 10

    # 1. Basic Implementation
    print("\n1. Basic Implementation:")
    dsu_basic = DSU_Basic(num_elements)
    dsu_basic.union(0, 1)
    dsu_basic.union(2, 3)
    dsu_basic.union(1, 3)  # Merge {0, 1} and {2, 3}
    print(f"  Find(0): {dsu_basic.find(0)}")  # Output: 3 (or 0, depending on which is parent)
    print(f"  Find(1): {dsu_basic.find(1)}")  # Output: 3 (or 0)
    print(f"  Find(2): {dsu_basic.find(2)}")  # Output: 3
    print(f"  Find(4): {dsu_basic.find(4)}")  # Output: 4
    print(f"  Parents: {dsu_basic.parent}")

    # 2. Path Compression
    print("\n2. Path Compression:")
    dsu_pc = DSU_PathCompression(num_elements)
    dsu_pc.union(0, 1)
    dsu_pc.union(2, 3)
    dsu_pc.union(1, 3)
    print(f"  Find(0): {dsu_pc.find(0)}")
    print(f"  Find(1): {dsu_pc.find(1)}")
    print(f"  Find(2): {dsu_pc.find(2)}")
    print(f"  Find(4): {dsu_pc.find(4)}")
    print(f"  Parents: {dsu_pc.parent}")

    # 3. Union by Rank
    print("\n3. Union by Rank:")
    dsu_ur = DSU_UnionByRank(num_elements)
    dsu_ur.union(0, 1)
    dsu_ur.union(2, 3)
    dsu_ur.union(1, 3)
    print(f"  Find(0): {dsu_ur.find(0)}")
    print(f"  Find(1): {dsu_ur.find(1)}")
    print(f"  Find(2): {dsu_ur.find(2)}")
    print(f"  Find(4): {dsu_ur.find(4)}")
    print(f"  Parents: {dsu_ur.parent}")
    print(f"  Ranks:   {dsu_ur.rank}")

    # 4. Path Compression + Union by Rank
    print("\n4. Path Compression + Union by Rank:")
    dsu_pcr = DSU_PathCompressionAndRank(num_elements)
    dsu_pcr.union(0, 1)
    dsu_pcr.union(2, 3)
    dsu_pcr.union(1, 3)
    print(f"  Find(0): {dsu_pcr.find(0)}")
    print(f"  Find(1): {dsu_pcr.find(1)}")
    print(f"  Find(2): {dsu_pcr.find(2)}")
    print(f"  Find(4): {dsu_pcr.find(4)}")
    print(f"  Parents: {dsu_pcr.parent}")
    print(f"  Ranks:   {dsu_pcr.rank}")

    # 5. DSU with List of Sets
    print("\n5. DSU with List of Sets:")
    dsu_sets = DSU_ListOfSets(num_elements)
    dsu_sets.union(0, 1)
    dsu_sets.union(2, 3)
    dsu_sets.union(1, 3)
    dsu_sets.union(4, 5)
    print(f"  Find(0): {dsu_sets.find(0)}")
    print(f"  Find(1): {dsu_sets.find(1)}")
    print(f"  Find(2): {dsu_sets.find(2)}")
    print(f"  Find(4): {dsu_sets.find(4)}")
    print(f"  Sets:    {dsu_sets.get_sets()}")

    # 6. Iterative Path Compression
    print("\n6. Iterative Path Compression:")
    dsu_ipc = DSU_IterativePathCompression(num_elements)
    dsu_ipc.union(0, 1)
    dsu_ipc.union(2, 3)
    dsu_ipc.union(1, 3)
    print(f"  Find(0): {dsu_ipc.find(0)}")
    print(f"  Find(1): {dsu_ipc.find(1)}")
    print(f"  Find(2): {dsu_ipc.find(2)}")
    print(f"  Find(4): {dsu_ipc.find(4)}")
    print(f"  Parents: {dsu_ipc.parent}")
    print(f"  Ranks:   {dsu_ipc.rank}")

    # Real-World Examples

    # 1. Network Connectivity
    print("\nReal-World Example 1: Network Connectivity")
    num_computers = 10
    network_dsu = DSU_PathCompressionAndRank(num_computers)
    connections = [(0, 1), (2, 3), (1, 3), (4, 5), (6, 7), (8, 9)]  # Computer connections
    for c1, c2 in connections:
        network_dsu.union(c1, c2)
    print(f"  Computers 0 and 3 connected? {network_dsu.find(0) == network_dsu.find(3)}")  # True
    print(f"  Computers 0 and 4 connected? {network_dsu.find(0) == network_dsu.find(4)}")  # False

    # 2. Social Network Connectivity
    print("\nReal-World Example 2: Social Network Connectivity")
    num_users = 8
    social_dsu = DSU_PathCompressionAndRank(num_users)
    friendships = [(0, 1), (1, 2), (3, 4), (5, 6), (4, 6)]
    for u1, u2 in friendships:
        social_dsu.union(u1, u2)
    print(f"  Users 0 and 2 connected? {social_dsu.find(0) == social_dsu.find(2)}")  # True
    print(f"  Users 0 and 5 connected? {social_dsu.find(0) == social_dsu.find(5)}")  # False
    print(f"  Users 3 and 5 connected? {social_dsu.find(3) == social_dsu.find(5)}") # True

    # 3. Image Processing (Segmentation) - Simplified Example
    print("\nReal-World Example 3: Image Processing (Segmentation)")
    # Imagine a 4x4 image represented as a 1D array (pixels 0-15)
    num_pixels = 16
    segment_dsu = DSU_PathCompressionAndRank(num_pixels)
    # Similar pixels are connected (example - based on some criteria)
    similar_pixels = [(0, 1), (1, 2), (4, 5), (5, 6), (8, 9), (9, 10), (12, 13), (13, 14),
                      (0, 4), (1, 5), (2, 6), (4, 8), (5, 9), (6, 10), (8, 12), (9, 13), (10, 14)]
    for p1, p2 in similar_pixels:
        segment_dsu.union(p1, p2)
    #  pixels 0,1,2,4,5,6,8,9,10,12,13,14 are in same segment. 3, 7, 11, 15 are in their own segments
    print(f"  Pixel 0 and 5 in same segment? {segment_dsu.find(0) == segment_dsu.find(5)}")  # True
    print(f"  Pixel 0 and 15 in same segment? {segment_dsu.find(0) == segment_dsu.find(15)}")  # False

    # 4. Kruskal's Algorithm (Minimum Spanning Tree) - Simplified
    print("\nReal-World Example 4: Kruskal's Algorithm (Minimum Spanning Tree) - Simplified")
    num_nodes = 7
    mst_dsu = DSU_PathCompressionAndRank(num_nodes)
    # Edges sorted by weight (simplified for demonstration)
    edges = [(0, 1), (1, 2), (0, 3), (3, 4), (2, 5), (4, 5), (5, 6)]  # (u, v) representing edges.  In real Kruskal, these would be (u,v, weight)
    mst_edges = [] # list to store edges in Minimum Spanning Tree
    for u, v in edges:
        if mst_dsu.find(u) != mst_dsu.find(v):  # If adding edge doesn't form a cycle
            mst_dsu.union(u, v)  # Add edge to MST
            mst_edges.append((u, v))
    print(f"  Edges in MST (simplified): {mst_edges}")  #  A possible MST: [(0, 1), (1, 2), (0, 3), (3, 4), (2, 5), (5, 6)]

    # 5. Circuit Network Analysis - Simplified
    print("\nReal-World Example 5: Circuit Network Analysis - Simplified")
    num_components = 8  #  Example:  Components in a circuit
    circuit_dsu = DSU_PathCompressionAndRank(num_components)
    connections = [(0, 1), (1, 2), (3, 4), (4, 5), (6, 7)]  # Connections between components
    for c1, c2 in connections:
        circuit_dsu.union(c1, c2)
    print(f"  Components 0 and 2 connected? {circuit_dsu.find(0) == circuit_dsu.find(2)}")  # True
    print(f"  Components 0 and 5 connected? {circuit_dsu.find(0) == circuit_dsu.find(5)}")  # False
    print(f"  Components 3 and 5 connected? {circuit_dsu.find(3) == circuit_dsu.find(5)}") # True

    # 6. Maze Generation - Simplified
    print("\nReal-World Example 6: Maze Generation - Simplified")
    width = 4
    height = 4
    num_cells = width * height
    maze_dsu = DSU_PathCompressionAndRank(num_cells)
    #  Walls are represented implicitly.  We'll just remove walls between adjacent cells
    #  in a way that doesn't create cycles.
    walls_to_remove = []
    #  For simplicity, consider only right and down walls.
    for r in range(height):
        for c in range(width):
            cell = r * width + c
            if c < width - 1: # Right wall
                right_neighbor = cell + 1
                walls_to_remove.append((cell, right_neighbor))
            if r < height - 1: # Down wall
                down_neighbor = cell + width
                walls_to_remove.append((cell, down_neighbor))
    import random
    random.shuffle(walls_to_remove) # Randomize wall removal order
    removed_walls = []
    for cell1, cell2 in walls_to_remove:
        if maze_dsu.find(cell1) != maze_dsu.find(cell2):
            maze_dsu.union(cell1, cell2)
            removed_walls.append((cell1, cell2))
    print(f"  Removed Walls (Simplified): {removed_walls}") #  List of walls removed to create maze.

if __name__ == "__main__":
    test_dsu_implementations()

File: C++/8-Graphs(BFS-DFS)/Algorithms/7-Union-Find(DisjointSetUnion-DSU)/Union-Find(DisjointSetUnion-DSU)Pro.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <stack>
#include <algorithm> // Added for std::reverse
#include <numeric>   // For iota

using namespace std;

// 1. Breadth-First Search (BFS) - Shortest Path in a Social Network
//
//   * Problem: Find the shortest path between two users in a social network.
//   * Application: Calculating degrees of separation (e.g., "Six Degrees of Kevin Bacon").
//   * Data Structure: Adjacency list for the graph, queue for BFS traversal.
void bfs(const vector<vector<int>>& graph, int start, int end, vector<int>& distances, vector<int>& previous) {
    int numNodes = graph.size();
    vector<bool> visited(numNodes, false);
    queue<int> q;

    // Initialize starting node
    visited[start] = true;
    distances[start] = 0;
    q.push(start);
    previous[start] = -1; // No previous node for the start node

    while (!q.empty()) {
        int current = q.front();
        q.pop();

        if (current == end) {
            break; // Stop if the target node is found
        }

        for (int neighbor : graph[current]) {
            if (!visited[neighbor]) {
                visited[neighbor] = true;
                distances[neighbor] = distances[current] + 1;
                previous[neighbor] = current; // Store the path
                q.push(neighbor);
            }
        }
    }
}

// Function to reconstruct the shortest path
vector<int> reconstructPath(int start, int end, const vector<int>& previous) {
    vector<int> path;
    int current = end;
    while (current != -1) {
        path.push_back(current);
        current = previous[current];
    }
    reverse(path.begin(), path.end()); // Reverse to get the correct order
    if (path[0] != start) {
        return {}; // Return empty path if no path exists.
    }
    return path;
}

void shortestPathSocialNetwork() {
    // Example graph representing a small social network
    //  0---1---2
    //  |   |
    //  3---4---5
    vector<vector<int>> graph = {
        {1, 3},    // 0
        {0, 2, 4}, // 1
        {1, 5},    // 2
        {0, 4},    // 3
        {1, 3, 5}, // 4
        {2, 4}     // 5
    };

    int numNodes = graph.size();
    vector<int> distances(numNodes, -1); // -1 indicates not visited
    vector<int> previous(numNodes, -1);
    int startUser = 0;
    int endUser = 5;

    bfs(graph, startUser, endUser, distances, previous);

    if (distances[endUser] != -1) {
        cout << "Shortest path between user " << startUser << " and " << endUser << ":\n";
        vector<int> path = reconstructPath(startUser, endUser, previous);
        for (int node : path) {
            cout << node << " ";
        }
        cout << "\nDistance: " << distances[endUser] << endl;
    } else {
        cout << "No path found between user " << startUser << " and " << endUser << endl;
    }
    cout << "--------------------------------------\n";
}

// 2. Depth-First Search (DFS) - Topological Sort for Task Scheduling
//
//   * Problem: Determine a valid order to execute tasks with dependencies.
//   * Application: Course scheduling, build systems (e.g., makefiles).
//   * Data Structure: Adjacency list for the directed graph, stack for DFS traversal.

void dfs(const vector<vector<int>>& graph, int start, vector<bool>& visited, stack<int>& stack) {
    visited[start] = true;
    for (int neighbor : graph[start]) {
        if (!visited[neighbor]) {
            dfs(graph, neighbor, visited, stack);
        }
    }
    stack.push(start); // Push the node onto the stack AFTER visiting its neighbors
}

void topologicalSort(const vector<vector<int>>& graph, vector<int>& result) {
    int numNodes = graph.size();
    vector<bool> visited(numNodes, false);
    stack<int> stack;

    for (int i = 0; i < numNodes; ++i) {
        if (!visited[i]) {
            dfs(graph, i, visited, stack);
        }
    }

    // Pop nodes from the stack to get the topological order
    while (!stack.empty()) {
        result.push_back(stack.top());
        stack.pop();
    }
}

void taskScheduling() {
    // Example graph representing task dependencies
    //  0 -> 1 -> 2
    //  |    |
    //  v    v
    //  3 -> 4
    vector<vector<int>> graph = {
        {1, 3}, // 0
        {2, 4}, // 1
        {},    // 2
        {4},    // 3
        {}     // 4
    };

    vector<int> sortedTasks;
    topologicalSort(graph, sortedTasks);

    cout << "Task scheduling order (Topological Sort): ";
    if (sortedTasks.size() == graph.size()) {
        for (int task : sortedTasks) {
            cout << task << " ";
        }
    } else {
        cout << "Cycle detected. No valid task order exists.";
    }

    cout << "\n--------------------------------------\n";
}

// 3. Union-Find (Disjoint Set Union) - Network Connectivity
//
//   * Problem: Determine if two computers are in the same network.
//   * Application: Network connection management, Kruskal's algorithm for Minimum Spanning Tree.
//   * Data Structure: Array to represent the parent of each node (node's representative).

class UnionFind {
public:
    vector<int> parent;
    vector<int> rank; // Optional: Used for optimization (union by rank)

    UnionFind(int n) : parent(n), rank(n, 0) {
        // Initialize each node as its own parent (singleton set)
        iota(parent.begin(), parent.end(), 0); // Sets parent[i] = i for all i
    }

    // Find the representative of the set that 'node' belongs to (with path compression)
    int find(int node) {
        if (node == parent[node]) {
            return node;
        }
        // Path compression: update the parent of 'node' to point directly to the root
        return parent[node] = find(parent[node]);
    }

    // Unite the sets containing 'node1' and 'node2'
    void unite(int node1, int node2) {
        int root1 = find(node1);
        int root2 = find(node2);

        if (root1 != root2) {
            // Union by rank (optional optimization)
            if (rank[root1] < rank[root2]) {
                parent[root1] = root2;
            } else if (rank[root1] > rank[root2]) {
                parent[root2] = root1;
            } else {
                parent[root2] = root1;
                rank[root1]++; // Increment rank if ranks are equal
            }
        }
    }

    // Check if two nodes are connected
    bool isConnected(int node1, int node2) {
        return find(node1) == find(node2);
    }
};

void networkConnectivity() {
    int numComputers = 6;
    UnionFind uf(numComputers);

    // Simulate network connections
    uf.unite(0, 1);
    uf.unite(2, 3);
    uf.unite(4, 5);
    uf.unite(1, 2); // Connects 0,1,2,3

    cout << "Network Connectivity:\n";
    cout << "Computer 0 and 3 are connected: " << (uf.isConnected(0, 3) ? "true" : "false") << endl; // true
    cout << "Computer 0 and 4 are connected: " << (uf.isConnected(0, 4) ? "true" : "false") << endl; // false
    cout << "Computer 2 and 5 are connected: " << (uf.isConnected(2, 5) ? "true" : "false") << endl; // false
    cout << "Computer 4 and 5 are connected: " << (uf.isConnected(4, 5) ? "true" : "false") << endl; // true
    cout << "--------------------------------------\n";
}

// 4.  BFS - Web Crawler
//
//    * Problem: Crawl a website to collect all accessible links.
//    * Application: Search engine indexing, data mining.
//    * Data Structure: Adjacency list (implicit, based on discovered links), queue for BFS.
//    * Simplification:  Simulating link discovery within a small, predefined set of pages.
void webCrawler() {
    // Simplified website representation (adjacency list)
    //  0: "home.html"
    //  1: "about.html"
    //  2: "services.html"
    //  3: "contact.html"
    //  4: "products.html"

    vector<vector<int>> website = {
        {1, 2},       // home.html links to about.html and services.html
        {0, 3},       // about.html links to home.html and contact.html
        {0, 4},       // services.html links to home.html and products.html
        {1},         // contact.html links to about.html
        {2}          // products.html links to services.html
    };
    vector<string> pageNames = {"home.html", "about.html", "services.html", "contact.html", "products.html"};
    int startPage = 0; // Start crawling from home.html
    int numPages = website.size();
    vector<bool> visited(numPages, false);
    queue<int> q;

    visited[startPage] = true;
    q.push(startPage);

    cout << "Web Crawler starting from " << pageNames[startPage] << ":\n";

    while (!q.empty()) {
        int currentPage = q.front();
        q.pop();

        cout << "Crawling page: " << pageNames[currentPage] << endl;

        for (int linkedPage : website[currentPage]) {
            if (!visited[linkedPage]) {
                visited[linkedPage] = true;
                q.push(linkedPage);
            }
        }
    }
    cout << "--------------------------------------\n";
}

// 5. DFS - Maze Solver
//
//   * Problem: Find a path from the start to the end of a maze.
//   * Application: Robotics navigation, game development.
//   * Data Structure: 2D vector (or array) to represent the maze, stack for DFS.

bool dfsMaze(vector<vector<int>>& maze, int row, int col, int endRow, int endCol, vector<pair<int, int>>& path) {
    int numRows = maze.size();
    int numCols = maze[0].size();

    // Base cases:
    if (row < 0 || row >= numRows || col < 0 || col >= numCols || maze[row][col] == 1) {
        return false; // Out of bounds or wall
    }
    if (row == endRow && col == endCol) {
        path.push_back({row, col});
        return true; // Found the end
    }

    maze[row][col] = 1; // Mark as visited (1 is wall, for simplicity)
    path.push_back({row, col});

    // Explore neighbors: up, down, left, right
    if (dfsMaze(maze, row - 1, col, endRow, endCol, path) ||
        dfsMaze(maze, row + 1, col, endRow, endCol, path) ||
        dfsMaze(maze, row, col - 1, endRow, endCol, path) ||
        dfsMaze(maze, row, col + 1, endRow, endCol, path)) {
        return true;
    }

    path.pop_back(); // Dead end, remove from path
    return false;
}

void mazeSolver() {
    // Example maze: 0 = path, 1 = wall
    vector<vector<int>> maze = {
        {0, 0, 0, 1, 0, 0, 0},
        {0, 1, 0, 1, 0, 1, 0},
        {0, 1, 0, 0, 0, 1, 0},
        {0, 1, 1, 1, 1, 1, 0},
        {0, 0, 0, 0, 0, 0, 0},
        {0, 1, 0, 1, 0, 1, 0},
        {0, 0, 0, 1, 0, 0, 0}
    };

    int startRow = 0;
    int startCol = 0;
    int endRow = 6;
    int endCol = 6;
    vector<pair<int, int>> path;

    if (dfsMaze(maze, startRow, startCol, endRow, endCol, path)) {
        cout << "Path found in the maze:\n";
        for (const auto& cell : path) {
            cout << "(" << cell.first << ", " << cell.second << ") -> ";
        }
        cout << "End\n";
    } else {
        cout << "No path found in the maze.\n";
    }
    cout << "--------------------------------------\n";
}

int main() {
    shortestPathSocialNetwork();
    taskScheduling();
    networkConnectivity();
    webCrawler();
    mazeSolver();

    return 0;
}
